File: 20-Valid-Parentheses.py
# Problem:
# Given a string of (, ), {, }, [, and ]
# Determing if the input string is valid

class Solution(object):
    def isValid(self, s):
        op = ['(','{','[']
        cl = [')','}',']']
        ongoing = []

        if s[0] in cl or s[-1] in op:
        	return(False)

        for index in range(3):
        	if s.count(op[index]) != s.count(cl[index]):
        		return(False)

        for para in s:
        	if para in op:
        		ongoing.append(para)
        	if para in cl:
        		prevPara = ongoing.pop()
        		for index in range(3):
        			if para == cl[index] and prevPara != op[index]:
        				return(False)

        return(True)

driver = Solution()
t1 = '()'
t2 = '({[]})'
t3 = '(]'
t4 = '([)]'
t5 = '}{'
t6 = '([{'
t7 = '([{}'

print(t1,driver.isValid(t1))
print(t2,driver.isValid(t2))
print(t3,driver.isValid(t3))
print(t4,driver.isValid(t4))
print(t5,driver.isValid(t5))
print(t6,driver.isValid(t6))
print(t7,driver.isValid(t7))

array = ['a','b','c','d']

array.reverse()
print(array.pop())
print(array.pop())

File: 20-Valid-Parentheses.cpp
// Problem:
// Given a string of (, ), {, }, [, and ]
// Determing if the input string uses valid (),[], and {}

class Solution {
public:
    bool isValid(string s) {

        std::stack<char> open_para;
        int len = s.length();

        for(int i = 0; i < len; i++)
        {
            if(s[i] == '(' || s[i] == '{' || s[i] == '[')
            {
                open_para.push(s[i]);
            }
            else
            {
                if(open_para.empty())
                {
                    return(false);
                }
                if(s[i] == ')' && open_para.top() != '(')
                {
                    return(false);
                }
                if(s[i] == '}' && open_para.top() != '{')
                {
                    return(false);
                }
                if(s[i] == ']' && open_para.top() != '[')
                {
                    return(false);
                }
                open_para.pop();
            }
        }
        if(open_para.empty())
        {
            return(true);
        }
        else
        {
            return(false);
        }
    }
};